### PR TITLE
Fix reported column number in ESLint

### PR DESCRIPTION
**Pull request by lydell**
## Before

screenshot from 2017-09-23 14-19-27

## After

screenshot from 2017-09-23 14-18-37

## Commit message

ESLint's column numbers are 0-based:
https://eslint.org/docs/developer-guide/working-with-rules#contextreport

> column - the 0-based column number at which the problem occurred.

Flow's column numbers are 1-based:

    $ cat index.js
    // @flow
    // The next line has an error at the first column: "abc" is not defined.
    abc

    $ flow check-contents < index.js --json --root index.js | jq '.errors[] | .message[] | .loc'
    {
      "source": "-",
      "type": "SourceFile",
      "start": {
        "line": 3,
        "column": 1,
        "offset": 82
      },
      "end": {
        "line": 3,
        "column": 3,
        "offset": 85
      }
    }

This commit simply subtracts the start column by 1.

Before:

    $ eslint index.js --format codeframe
    error: identifier `abc`:  Could not resolve name (flowtype-errors/show-errors) at index.js:3:2:
      1 | // @flow
      2 | // The next line has an error at the first column: "abc" is not defined.
    > 3 | abc
        |  ^
      4 |

    1 error found.

After:

    $ eslint index.js --format codeframe
    error: identifier `abc`:  Could not resolve name (flowtype-errors/show-errors) at index.js:3:1:
      1 | // @flow
      2 | // The next line has an error at the first column: "abc" is not defined.
    > 3 | abc
        | ^
      4 |

    1 error found.

Shouldn't the end column be subtracted by 1 as well? It seems like
ESLint's end column is exclusive, while Flow's is inclusive. So the
end column should stay unchanged. I tested this by using the
linter-eslint plugin in the Atom editor. When subtracting 1 from the end
column only "ab" of "abc" in index.js above is underlined as an error.
When keeping the end column the entire word "abc" is underlined (which
is the desired result). Here you can see what it looks like if 1 is subtracted from the end as well:

screenshot from 2017-09-23 14-23-13

I have consistently noted that the error locations are off-by-one after
having used eslint-plugin-flowtype-errors for several months, should it
should be safe to always simply subtract 1 from the start column.

### CODE REVIEW

**Comment by lydell** [comment]
Travis fails because of `ESLint couldn't find the plugin "eslint-plugin-flowtype-errors". ` That does not feel related to this PR?

**Comment by lydell** [comment]
**TL;DR** Please tell me if there's anything you'd like me to change before merging! :)

---

> Maybe we could do `loc.start.column--` directly in the `src/collect.js` file.

Nope :)

```
❯ git diff
diff --git a/src/collect.js b/src/collect.js
index 3cd31f7..71f6c76 100644
--- a/src/collect.js
+++ b/src/collect.js
@@ -242,6 +242,10 @@ export function collect(
       const loc = firstMessage.loc;
       const finalMessage = entireMessage.replace(/\.$/, '');
 
+      if (firstMessage.loc) {
+        firstMessage.loc.start.column--;
+      }
+
       return {
         ...(process.env.DEBUG_FLOWTYPE_ERRRORS === 'true'
           ? parsedJSONArray

~/forks/eslint-plugin-flowtype-errors fix-column-numbers * 
❯ yarn lint
yarn lint v1.0.1
$ eslint --cache --format=node_modules/eslint-formatter-pretty .

  src/collect.js:246:9
  ✖  246:9  Unary operator -- used.  no-plusplus

  1 error

error Command failed with exit code 1.
```

We _could_ move the current non-mutating column-adjusting code to `src/collect.js`, but as far as I understand that file it only deals with Flow stuff. So, to me, it didn't feel right to add the ESLint specific column adjustment there. Your call.

---

I'm running yarn 1.0.1 on Ubuntu 16.04, and yarn actually fails to install eslint-plugin-flowtype-errors for me too (even though it is a [dependency of eslint-config-bliss](https://github.com/amilajack/eslint-config-bliss/blob/8562b76f40b132eae50e51453a75e22aa75dc725/package.json#L32)). I have to manually do `yarn add eslint-plugin-flowtype-errors` (and `git checkout .` to get rid of the changes in package.json and yarn.lock) to be able to run the tests (lint).

<details>
<summary>Console session</summary>

<pre>~/forks/eslint-plugin-flowtype-errors fix-column-numbers 
❯ rm node_modules/ -r

~/forks/eslint-plugin-flowtype-errors fix-column-numbers 
❯ yarn
yarn install v1.0.1
[1/5] Validating package.json...
[2/5] Resolving packages...
[3/5] Fetching packages...
info fsevents@1.1.1: The platform "linux" is incompatible with this module.
info "fsevents@1.1.1" is an optional dependency and failed compatibility check. Excluding it from installation.
[4/5] Linking dependencies...
warning "ajv-keywords@1.5.1" has incorrect peer dependency "ajv@>=4.10.0".
warning "eslint-plugin-prettier@2.1.2" has incorrect peer dependency "prettier@>= 0.11.0".
[5/5] Building fresh packages...
Done in 5.55s.

~/forks/eslint-plugin-flowtype-errors 6s fix-column-numbers 
❯ yarn test
yarn test v1.0.1
$ flow && npm run lint && npm run spec
No errors!

> eslint-plugin-flowtype-errors@3.3.1 lint /home/lydell/forks/eslint-plugin-flowtype-errors
> eslint --cache --format=node_modules/eslint-formatter-pretty .


Oops! Something went wrong! :(

ESLint couldn't find the plugin "eslint-plugin-flowtype-errors". This can happen for a couple different reasons:

1. If ESLint is installed globally, then make sure eslint-plugin-flowtype-errors is also installed globally. A globally-installed ESLint cannot find a locally-installed plugin.

2. If ESLint is installed locally, then it's likely that the plugin isn't installed correctly. Try reinstalling by running the following:

    npm i eslint-plugin-flowtype-errors@latest --save-dev

If you still can't figure out the problem, please stop by https://gitter.im/eslint/eslint to chat with the team.

npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! eslint-plugin-flowtype-errors@3.3.1 lint: `eslint --cache --format=node_modules/eslint-formatter-pretty .`
npm ERR! Exit status 1
npm ERR! 
npm ERR! Failed at the eslint-plugin-flowtype-errors@3.3.1 lint script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     /home/lydell/.npm/_logs/2017-09-24T07_08_18_335Z-debug.log
error Command failed with exit code 1.

~/forks/eslint-plugin-flowtype-errors fix-column-numbers 
❯ yarn add eslint-plugin-flowtype-errors
yarn add v1.0.1
[1/5] Validating package.json...
[2/5] Resolving packages...
[3/5] Fetching packages...
info fsevents@1.1.1: The platform "linux" is incompatible with this module.
info "fsevents@1.1.1" is an optional dependency and failed compatibility check. Excluding it from installation.
[4/5] Linking dependencies...
warning "ajv-keywords@1.5.1" has incorrect peer dependency "ajv@>=4.10.0".
warning "eslint-plugin-prettier@2.1.2" has incorrect peer dependency "prettier@>= 0.11.0".
[5/5] Building fresh packages...
success Saved lockfile.
success Saved 1 new dependency.
└─ eslint-plugin-flowtype-errors@3.3.1
Done in 2.61s.

~/forks/eslint-plugin-flowtype-errors fix-column-numbers * 
❯ git status
On branch fix-column-numbers
Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git checkout -- <file>..." to discard changes in working directory)

	modified:   package.json
	modified:   yarn.lock

no changes added to commit (use "git add" and/or "git commit -a")

~/forks/eslint-plugin-flowtype-errors fix-column-numbers * 
❯ git checkout .

~/forks/eslint-plugin-flowtype-errors fix-column-numbers 
❯ yarn test
yarn test v1.0.1
$ flow && npm run lint && npm run spec
No errors!

> eslint-plugin-flowtype-errors@3.3.1 lint /home/lydell/forks/eslint-plugin-flowtype-errors
> eslint --cache --format=node_modules/eslint-formatter-pretty .


> eslint-plugin-flowtype-errors@3.3.1 spec /home/lydell/forks/eslint-plugin-flowtype-errors
> npm run build && jest test/format.spec.js


> eslint-plugin-flowtype-errors@3.3.1 build /home/lydell/forks/eslint-plugin-flowtype-errors
> cross-env NODE_ENV=production babel src --out-dir dist

src/collect.js -> dist/collect.js
src/format.js -> dist/format.js
src/index.js -> dist/index.js
 PASS  test/format.spec.js (24.679s)
  Format
    ✓ 1.example.js - should have expected properties (11ms)
    ✓ 2.example.js - should have expected properties (2ms)
    ✓ 3.example.js - should have expected properties (8ms)
    ✓ 4.example.js - should have expected properties (2ms)
    ✓ 5.example.js - should have expected properties (1ms)
    ✓ 6.example.js - should have expected properties (1ms)
    ✓ 7.example.js - should have expected properties (1ms)
    ✓ 8.example.js - should have expected properties (2ms)
    ✓ 9.example.js - should have expected properties (2ms)
  Check codebases
    ✓ flow-pragma-1 - eslint should give expected output (1640ms)
    ✓ flow-pragma-2 - eslint should give expected output (2618ms)
    ✓ no-flow-pragma - eslint should give expected output (629ms)
    ✓ project-1 - eslint should give expected output (2726ms)
    ✓ run-all - eslint should give expected output (2451ms)
    ✓ run-all-flowdir - eslint should give expected output (2442ms)
    ✓ coverage-ok - eslint should give expected output (2587ms)
    ✓ coverage-ok2 - eslint should give expected output (2582ms)
    ✓ coverage-fail - eslint should give expected output (2576ms)
    ✓ coverage-fail2 - eslint should give expected output (2630ms)
    ✓ folder with spaces - eslint should give expected output (483ms)

Test Suites: 1 passed, 1 total
Tests:       20 passed, 20 total
Snapshots:   20 passed, 20 total
Time:        24.71s, estimated 25s
Ran all test suites matching "test/format.spec.js".
Done in 29.77s.</pre>

</details>

**Comment by lydell** [comment]
I've reported the yarn issue here: https://github.com/yarnpkg/yarn/issues/4535

**Comment by jdmota** [comment]
I don't have a strong opinion on this. Like I said, the PR is good :)

I just thought that decrementing was easier to read, and that the `collect` function could return the errors already normalized for Eslint.

@amilajack What do you think?

**Comment by jdmota** [comment]
@lydell Btw, maybe you also could update the changelog?

**Comment by jdmota** [comment]
@lydell I think the changelog was outdated. Can you do something like:

```
## Master

### Fixed

- Fixed off-by-one error in reported column numbers.

## v3.3.0

### Added

// etc...
```

Sorry about that.

**Comment by amilajack** [comment]
@lydell thanks for the PR and @jdmota thanks for reviewing!

**Comment by amilajack** [comment]
Just pushed `3.3.2` 🎉

**Comment by lydell** [comment]
Thanks! You might want to change "Master" to "3.3.2" in the changelog :)
